sound/soc/sof/pcm.c

```diff
@@ -214,6 +214,9 @@ static int sof_pcm_hw_params(struct snd_pcm_substream *substream,
 	INIT_WORK(&spcm->stream[substream->stream].period_elapsed_work,
 		  sof_pcm_period_elapsed_work);
 
+	/* clear hw_params_upon_resume flag */
+	spcm->hw_params_upon_resume[substream->stream] = 0;
+
```

Comment:
this is already reset during the sof_pcm_open() step. Why does it need to be reset *again* in hw_params?
The commit message is really far from clear, did you mean that prepare would call sof_pcm_hw_params a second time? multiple means two and more, so not sure what you are trying to say.

Reply:
When there is a xrun, ALSA will call the prepare(). So please image the below flow (just a case to reproduce the situation of call prepare()):
(1) aplay -Dhw:0,0 -f dat wav/48California_Dreaming.wav -vv -i
(2) Pause music
(3) Open another terminal, run 'sudo rtcwake -m mem -s 2' to enter S3 and wake up
(4) Resume music
(5) Ctl+z to put background
(6) 'fg'. failed and error seen now.

In this case, prepare() will be called in step "(4) Resume music". However in "step (6)" there is a xrun, which will cause alsa core to call prepare(). In this case, you will find the hw_params_upon_resume flag is still active, which will cause hw_params called twice. And then the FW will fail to handle such situation.

Reply:
@plbossart Or I can add the xrun case in the commit description. What do you think of it?

Reply:
we also have 2 fields called spcm->hw_params_upon_resume and hda_stream->hw_params_upon_resume, and I wonder if the behavior needs to be similar for both? You'd want to clarify for reviewers if the logic is the same on purpose?

Reply:
hda_stream->hw_params_upon_resume seems good. Ranjani has already cleared that flag in hw_params. The only difference is hda_stream->hw_params_upon_resume is not cleared in open. It is OK to me to clear or not to clear the flag in open. To make the logic be the same, I will remove the clear from pcm open.

> we also have 2 fields called spcm->hw_params_upon_resume and hda_stream->hw_params_upon_resume, and I wonder if the behavior needs to be similar for both? You'd want to clarify for reviewers if the logic is the same on purpose?

Reply:
@plbossart we now have different flags for the same purpose for the BE and FE. The original intention to not clear the FW flag in hw_params was to re-use the same flag in the BE. But after we added a separate flags, it makes sense to clear it in hw_params.